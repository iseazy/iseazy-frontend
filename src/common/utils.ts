export function getFormattedTimeDifference(fromDate: number, toDate: number = Date.now()): string {
    const millis = toDate -fromDate;
    const minutes = Math.floor(millis / 60000);
    const seconds = ((millis % 60000) / 1000).toFixed(0);
    let gameDuration = '';
    
    if (minutes > 0) {
        gameDuration += `${minutes}min `;
    }

    return `${gameDuration}${seconds}s`;
}

export function shuffleList<T>(items: Array<T>): Array<T> {
    return items.sort(() => Math.random() - 0.5);
}
