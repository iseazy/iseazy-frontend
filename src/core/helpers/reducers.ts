import { Card } from "../../hooks/AppContext";


export const updateFlippedCardsArray = (array: Card[], value: Card) => {
    const existsAlready = array.some((item) => item.id === value.id)
    if(existsAlready) return array;
    return [...array,value];
}