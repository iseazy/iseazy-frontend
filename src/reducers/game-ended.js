import { GAME_STATE } from '../constants';

export const gameEnded = (state) => ({
    ...state,
    gameState: GAME_STATE.END,
    totalTime: (Date.now() - state.startTime)/1000
});
