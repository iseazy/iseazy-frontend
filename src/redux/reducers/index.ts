import { combineReducers } from "redux";
import gameReducer from "./gameReducer";

export const rootReducer = combineReducers({
  gameReducer,
});

export type RootState = ReturnType<typeof rootReducer>