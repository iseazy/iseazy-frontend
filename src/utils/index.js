export {
    assignCardId,
    createCard,
    flipCardById,
    isCardClicked,
    isCardFlipped,
    isCardNotClicked,
    hasCardSharedId,
    unclickCard,
    unflipClickedCard
} from './card';

export {
    areClickedCardsPaired,
    createDeck,
    isDeckSolved
} from './deck';

export { isGameEnd, isGameIdle } from './game';

export { shuffleArray } from './suffle-array';

export { getFormattedTime } from './time';
