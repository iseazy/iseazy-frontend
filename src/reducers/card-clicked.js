import { REPEATED_CARDS } from '../constants';
import { flipCardById, isCardClicked, isCardFlipped, isCardNotClicked } from '../utils';

export const cardClicked = (state, id) => { 
    const { cards } = state;
    const clickedCard = cards.find(card => card.id === id);

    let updatedCards = cards;

    if ([cards.filter(isCardClicked).length < REPEATED_CARDS,
        !isCardFlipped(clickedCard),
        isCardNotClicked(clickedCard)].every(Boolean)) {
        updatedCards = cards.map(flipCardById(id))
    }

    return {
        ...state,
        cards: updatedCards
    } 
};