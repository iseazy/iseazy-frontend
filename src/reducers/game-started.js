import { CARDS_SRC, GAME_STATE } from '../constants';
import { createDeck } from '../utils';

export const gameStarted = (state) => ({
    ...state,
    gameState: GAME_STATE.PLAY,
    cards: createDeck(CARDS_SRC),
    startTime: Date.now()
});
