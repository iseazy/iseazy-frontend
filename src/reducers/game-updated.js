import { REPEATED_CARDS } from '../constants';
import { areClickedCardsPaired, isCardClicked, unclickCard, unflipClickedCard } from '../utils';

export const gameUpdated = (state) => {
    const { cards } = state;
    const clickedCards = cards.filter(isCardClicked);

    let updatedCards = cards;

    if (clickedCards.length >= REPEATED_CARDS) {
        updatedCards = areClickedCardsPaired(clickedCards)
            ? cards.map(unclickCard)
            : cards.map(unflipClickedCard);
    }

    return {
        ...state,
        cards: updatedCards
    }
};
