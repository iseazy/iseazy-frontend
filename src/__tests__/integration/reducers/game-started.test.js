import { gameStarted } from '../../../reducers/game-started';
import { createDeck } from '../../../utils';

jest.mock('../../../utils', () => ({
    createDeck: jest.fn()
}));

describe('gameStarted()', () => {
    beforeAll(() => {
        createDeck.mockReturnValue(mockDeck);
    });

    it('should initialize gameState, create shuffled deck and set startTime', () => {
        Date.now = jest.fn(() => 123);
        expect(gameStarted({})).toMatchSnapshot();
    });
});
