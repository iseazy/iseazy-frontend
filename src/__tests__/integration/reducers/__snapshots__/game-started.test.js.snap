// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gameStarted() should initialize gameState, create shuffled deck and set startTime 1`] = `
Object {
  "cards": undefined,
  "gameState": "play",
  "startTime": 123,
}
`;
